Cleaned up template loader tests.

prestontimmons opened the pull request:
If I can be entertained here, I'd like to clean up ``template_tests/test_loaders.py``. I have a number of tests I want to add to this file from work on #3475. This cleanup makes that a bit nicer.

Some explanation on a few of the changes:

1) I modified the egg loader tests to clean up after themselves. They were modifying sys.modules without putting the changes back.

2) I moved ``PriorityLoader`` and ``PriorityCachedLoader`` into ``GetTemplateTests``. I'll be adding additional tests to this test group as well.

3) I used ``get_template`` in favor of ``loader.get_template``, etc.

4) I removed the ``test_empty_egg`` egg test. That to me was redundant with ``test_non_existing``.

The rest of the changes are mainly cleaning up docstrings and indentation.

aaugustin commented:
Great! I'll take a look at this tonight (CET).

aaugustin commented on tests/template_tests/test_loaders.py:
```diff
@@ -126,18 +130,60 @@ def test_templatedir_caching(self):
+        )
+
+
+class GetTemplateTests(SimpleTestCase):
```
If you were testing the `get_template` function then the tests would go in the `template_loader` test app. Can you give this class another name (or move it)?

By the way, the tests would be much cleaner if they created a suitable `Engine` in `setUp` and ran the tests on that engine instead of the global one, like `template_tests.utils.setup` does.

I don't want to send you down a rabbit hole when you were just doing some light cleanup :-) but that's something to keep in mind if you add new tests.

aaugustin commented on tests/template_tests/test_loaders.py:
```diff
@@ -145,22 +191,20 @@ def test_missing_template_is_cached(self):
     'DIRS': [TEMPLATES_DIR],
 }])
 class RenderToStringTest(SimpleTestCase):
```
Many of these tests look redundant with the new (and more complete) tests I added in the `template_loader` test app. What about removing them?

One thing that's missing from the new tests is passing a context in argument.

aaugustin commented on tests/template_tests/test_loaders.py:
```diff
@@ -145,22 +191,20 @@ def test_missing_template_is_cached(self):
-            'No template names provided$',
-            loader.select_template, [])
+        with self.assertRaisesMessage(TemplateDoesNotExist, 'No template names provided'):
+            select_template([])
```
Testing `select_template` in `RenderToStringTest`, grr...

aaugustin commented:
Thanks a lot for tackling this cleanup, it's really needed. You're on the right track but I think you could take it a bit further. Let me know if you'd like to improve the PR. Otherwise I'll merge it as it's already a worthwhile improvement in its current state.

prestontimmons commented:
Thanks for the feedback. I'll definitely update the things you mentioned.

LOL on the ``select_template``. This isn't the only thing in template_tests that's out of place or redundant.

prestontimmons commented:
Okay. I've updated the test cleanup. This now also cleans up ``tests.py``, which before was a haphazard collection of tests. I tried to keep the subsequent commits small so their easy to review.

The tests no longer call anything from ``django.template.loader`` anymore, besides one place. It's still used by the inclusion tag file. This can probably be fixed, too. I'll look at that later.

I removed the redundant ``render_to_string`` tests and moved the test passing a context to ``DeprecatedRenderToStringTests``. I think that test is appropriate there. Do we want to add it to the tests in ``template_loaders`` as well?

Finally, I plan to update the filesystem and app loader security tests to not use the deprecated dirs argument. That will be easier with the next commit I want to add after this cleanup.

aaugustin commented:
Excellent work! If you address my comments, I can merge it, unless you want to make the updates you talk about in your last comment first.

prestontimmons commented:
Thanks for reviewing this. I think I'll address the left over global settings dependencies in another PR.

aaugustin commented:
I rebased and merged the branch. Thanks!
